src/hotspot/share/gc/shenandoah/shenandoahEvacOOMHandler.cpp

```diff
@@ -31,22 +31,46 @@
 
 const jint ShenandoahEvacOOMHandler::OOM_MARKER_MASK = 0x80000000;
 
-ShenandoahEvacOOMHandler::ShenandoahEvacOOMHandler() :
-  _threads_in_evac(0) {
+ShenandoahEvacOOMHandler::ShenandoahEvacOOMHandler() {
+  for (int i = 0; i < EVAC_COUNTER_BUCKETS; i++)
+    _threads_in_evac[i].bits = 0;
 }
 
-void ShenandoahEvacOOMHandler::wait_for_no_evac_threads() {
-  while ((Atomic::load_acquire(&_threads_in_evac) & ~OOM_MARKER_MASK) != 0) {
+volatile jint *ShenandoahEvacOOMHandler::threads_in_evac_ptr(Thread* t) {
+  uint64_t key = (uintptr_t)t;
+  key ^= (key >> 33);
+  key *= UINT64_C(0xff51afd7ed558ccd);
+  key ^= (key >> 33);
+  key *= UINT64_C(0xc4ceb9fe1a85ec53);
+  key ^= (key >> 33);
+
+  return &_threads_in_evac[key % EVAC_COUNTER_BUCKETS].bits;
+}
+
+void ShenandoahEvacOOMHandler::wait_for_one_counter(volatile jint *ptr) {
+  // We might be racing against handle_out_of_memory_during_evacuation()
+  // setting the OOM_MARKER_MASK bit so we must make sure it is set here
+  // *and* the counter is zero.
+  while (Atomic::load_acquire(ptr) != OOM_MARKER_MASK) {
     os::naked_short_sleep(1);
```

Comment:
Not sure if SpinPause() may be better here? @shipilev probably knows more.

Reply:
I think we'd probably want some back-off here rather than spinning indefinitely? E.g. spin N times and then start sleeping.